### ISSUE TITLE
Connection lost after resuming from hibernate

### ISSUE THREAD

**Issue by ElStupid** [OpenVPN/ovpn-dco-win#64]
To be fair, I'm a fairly new user of OpenVPN, using it as my default connection. The OpenVPN-Gui was supplied by Surfshark, but their version did not auto connect on startup of my system. So I found v11.46.0.0, which has an option to Enable auto restart of active connections. 

However, after resuming my machine from hibernate, inertnet connection is lost and doesn't auto reconnect. The tray icon seems happily green, but I have to reconnect myself in order t have internet connection again, which seems like unintended behaviour. 
Is this a known issue, which will be fixed in an upcoming release?

Thnx!

**Comment by lstipakov** [comment]
I can confirm that. With hibernate (S4) it is easy to reproduce, and with "modern standby" (S0 low power), which is default on my system, it depends - usually when my system wakes up from "modern standby" it immediately got "keepalive timeout" notification in userspace, even though according to kernel logs it was fired some time ago.

During hibernate and, to some extent, modern standby, "relative" timers (the ones I use) are not ticking, so when system wakes up timers are continued where they are left off, which makes keepalive timeout experience sub-optimal.

I'll refactor timers implementation and use instead single timer with 1sec resolution, where I compare `last` and `now` values obtained with `KeQuerySystemTime` call. This is similar to what OpenVPN2 userspace is doing.

**Comment by lstipakov** [comment]
Here is a new version of the driver with refactored timers implementation. I have tested both hibernate and "modern standby" - in both cases the client reconnected as soon as machine waked up. @selvanair @ElStupid if you could test this as well it would be great. Just pick the driver depends on your OS and run (I also added devcon.exe to the archive):

```
PS C:\Users\lev\Projects\ovpn-dco-win\signed\1.1.0\win11> .\devcon.exe install .\ovpn-dco.inf ovpn-dco
Device node created. Install is complete when drivers are installed...
Updating drivers for ovpn-dco from C:\Users\lev\Projects\ovpn-dco-win\signed\1.1.0\win11\ovpn-dco.inf.
Drivers installed successfully.
```

The driver is attestation signed so you should be able to install it without any additional steps. I have bumped the version so it should update the driver for existing devices - you should be seeing it from openvpn log:

```
Thu Apr 11 23:12:59 2024 OpenVPN 2.6.10 [git:v2.6.10/ba0f62fb950c56a0] Windows [SSL (OpenSSL)] [LZO] [LZ4] [PKCS11] [AEAD] [DCO] built on Mar 20 2024
Thu Apr 11 23:12:59 2024 Windows version 10.0 (Windows 10 or greater), amd64 executable
Thu Apr 11 23:12:59 2024 library versions: OpenSSL 3.2.1 30 Jan 2024, LZO 2.10
Thu Apr 11 23:12:59 2024 DCO version: 1.1.0
```

[ovpn-dco-win-1.1.0-win10.zip](https://github.com/OpenVPN/ovpn-dco-win/files/14951055/ovpn-dco-win-1.1.0-win10.zip)
[ovpn-dco-win-1.1.0-win11.zip](https://github.com/OpenVPN/ovpn-dco-win/files/14951056/ovpn-dco-win-1.1.0-win11.zip)

**Comment by selvanair** [comment]
This works as expected. On resume from suspend to disk (aka hybernate / S4), ping-restart triggered almost right away.

Interestingly, I thought `timerConfig.TolerableDelay = TolerableDelayUnlimited;` in the patch might break the tunnel in "Modern standby with network" mode where the tunnel used to stay up earlier. But it does continue to work. Hope that's not because my machine was only lightly loaded.

**Comment by ElStupid** [comment]
> Here is a new version of the driver with refactored timers implementation. I have tested both hibernate and "modern standby" - in both cases the client reconnected as soon as machine waked up. @selvanair @ElStupid if you could test this as well it would be great. Just pick the driver depends on your OS and run (I also added devcon.exe to the archive):
> 

Thanks! Will try later today and keep you posted.

**Comment by ElStupid** [comment]
> Note: Just installing the driver using devcon did not work for me as OpenVPN.exe kept reporting DCO version as 1.0.1 and the ping-restart behaviour remained as before. Although the driver tab of the adapter showed 1.1.0. Deleting all existing dco adapters and recreating using tapctl.exe fixed that.

A little beyond my technical knowledge, but we'll see ;)

**Comment by lstipakov** [comment]
@ElStupid The easiest way would probably be open `Device Manager`, expand `Network Adapters`, find `OpenVPN Data Channel Offload`, right click -> `Remove Device`, and in an popped up dialog box select checkbox `Remove Drivers`. After that you can:

```
PS C:\Users\lev\Projects\ovpn-dco-win\signed\1.1.0\win11> .\devcon.exe install .\ovpn-dco.inf ovpn-dco
Device node created. Install is complete when drivers are installed...
Updating drivers for ovpn-dco from C:\Users\lev\Projects\ovpn-dco-win\signed\1.1.0\win11\ovpn-dco.inf.
Drivers installed successfully.
```

**Comment by lstipakov** [comment]
Just found (and fixed) a bug in 1.1.0. I'll attach the client installer here soonish, which includes the new version of the driver. (1.1.1)

**Comment by ElStupid** [comment]
Just tried this and works flawlessly, thanks so much. After resuming from hibernate I'm instantly connected now.

**Comment by selvanair** [comment]
The change from 1.1.0 to 1.1.1 would only affect TCP instances right? I do not have one at hand to test against, but looks good to me.

**Comment by lstipakov** [comment]
Correct, TCP only. You can get the new installer for different platforms here [https://github.com/OpenVPN/openvpn-build/actions/runs/8685747402](https://github.com/OpenVPN/openvpn-build/actions/runs/8685747402)

**Comment by lstipakov** [comment]
Fixed in 2.6.10-I002.
